File: flutter-client/ios/Runner/AudioOutputProfileRegistry.swift
import AVFoundation

/// iOS audio output profile registry.
/// Maps profile IDs to AVAudioSession category/mode/options.
/// IDs match Android [AudioOutputProfileRegistry] for shared [AudioOutputProfileStore] keys.
enum AudioOutputProfileRegistry {
  struct Option {
    let id: String
    let title: String
    let mode: AVAudioSession.Mode
    let defaultToSpeaker: Bool
    let allowBluetoothA2dp: Bool
    let duckOthers: Bool
  }

  static let options: [Option] = [
    Option(
      id: ID_MEDIA,
      title: ID_MEDIA,
      mode: .default,
      defaultToSpeaker: true,
      allowBluetoothA2dp: false,
      duckOthers: false
    ),
    Option(
      id: ID_VOICE_CALL,
      title: ID_VOICE_CALL,
      mode: .voiceChat,
      defaultToSpeaker: false,
      allowBluetoothA2dp: false,
      duckOthers: false
    ),
    Option(
      id: ID_VOICE_CALL_BT,
      title: ID_VOICE_CALL_BT,
      mode: .voiceChat,
      defaultToSpeaker: false,
      allowBluetoothA2dp: true,
      duckOthers: false
    ),
    Option(
      id: ID_GAME,
      title: ID_GAME,
      mode: .gameChat,
      defaultToSpeaker: true,
      allowBluetoothA2dp: false,
      duckOthers: false
    ),
    Option(
      id: ID_RAW,
      title: ID_RAW,
      mode: .measurement,
      defaultToSpeaker: true,
      allowBluetoothA2dp: false,
      duckOthers: false
    ),
    Option(
      id: ID_NOTIFICATION,
      title: ID_NOTIFICATION,
      mode: .voiceChat,
      defaultToSpeaker: true,
      allowBluetoothA2dp: false,
      duckOthers: true
    ),
  ]

  static func option(for id: String) -> Option {
    options.first { $0.id == id }
      ?? options.first { $0.id == ID_MEDIA }
      ?? options[0]
  }

  static func applySession(profileId: String) throws {
    let profile = option(for: profileId)
    var options: AVAudioSession.CategoryOptions = [.allowBluetooth]
    if profile.defaultToSpeaker {
      options.insert(.defaultToSpeaker)
    }
    if profile.allowBluetoothA2dp {
      options.insert(.allowBluetoothA2DP)
    }
    if profile.duckOthers {
      options.insert(.duckOthers)
    }
    let session = AVAudioSession.sharedInstance()
    try session.setCategory(.playAndRecord, mode: profile.mode, options: options)
    try session.setActive(true)
  }

  static let ID_MEDIA = "media"
  static let ID_VOICE_CALL = "voice_call"
  static let ID_VOICE_CALL_BT = "voice_call_bt"
  static let ID_GAME = "game"
  static let ID_RAW = "raw"
  static let ID_NOTIFICATION = "notification"
}

File: flutter-client/android/app/src/main/kotlin/ru/outsidepro_arts/owalkie/flutter/AudioOutputProfileRegistry.kt
package ru.outsidepro_arts.owalkie.flutter

import android.content.Context
import android.media.AudioManager

/**
 * Audio output profile registry for Android.
 * Maps profile IDs to AudioManager mode + BT SCO routing.
 * AAudio usage/contentType and capture usage are handled via FFI in Dart.
 */
object AudioOutputProfileRegistry {
    data class Option(
        val id: String,
        val title: String,
        val audioManagerMode: Int,
        val enableBtSco: Boolean,
    )

    fun listOptions(context: Context): List<Option> {
        return listOf(
            Option(ID_MEDIA, context.getString(R.string.audio_output_media), MODE_NORMAL, false),
            Option(ID_VOICE_CALL, context.getString(R.string.audio_output_voice_call), MODE_IN_COMMUNICATION, false),
            Option(ID_VOICE_CALL_BT, context.getString(R.string.audio_output_voice_call_bt), MODE_IN_COMMUNICATION, true),
            Option(ID_GAME, context.getString(R.string.audio_output_game), MODE_NORMAL, false),
            Option(ID_RAW, context.getString(R.string.audio_output_raw), MODE_NORMAL, false),
            Option(ID_NOTIFICATION, context.getString(R.string.audio_output_notification), MODE_NORMAL, false),
        )
    }

    fun applyProfile(context: Context, id: String) {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        when (id) {
            ID_VOICE_CALL -> {
                audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
                @Suppress("DEPRECATION")
                audioManager.isSpeakerphoneOn = true
                AudioRouteHelper.disableBluetoothInputRoute(audioManager)
            }
            ID_VOICE_CALL_BT -> {
                audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
                @Suppress("DEPRECATION")
                audioManager.isSpeakerphoneOn = false
                AudioRouteHelper.enableBluetoothInputRoute(audioManager)
            }
            else -> {
                audioManager.mode = AudioManager.MODE_NORMAL
                @Suppress("DEPRECATION")
                audioManager.isSpeakerphoneOn = false
                AudioRouteHelper.disableBluetoothInputRoute(audioManager)
            }
        }
    }

    const val ID_MEDIA = "media"
    const val ID_VOICE_CALL = "voice_call"
    const val ID_VOICE_CALL_BT = "voice_call_bt"
    const val ID_GAME = "game"
    const val ID_RAW = "raw"
    const val ID_NOTIFICATION = "notification"

    private const val MODE_NORMAL = 0
    private const val MODE_IN_COMMUNICATION = 3
}